### PR TITLE
Add ability to read tif files

### PR DESCRIPTION
**Pull request by csherwood-usgs**
Merging from crs_dev. See issue #30

### CODE REVIEW

File: doodler/components.py [hunk 1]
```diff
@@ -39,7 +42,7 @@ class Toggle(pn.reactive.ReactiveHTML):
     color = param.String(doc='Color of the border in hex format')
 
     _template = """
-    <button id="button" style="text-decoration: {{ 'underline' if active else 'normal' }};border-color:{{ color }};border-width:4px;border-radius:5%;padding:10px;font-weight:{{ 'bold' if active else 'normal' }}" onclick="${_update}">
```
Comment (maximlt) [inline doodler/components.py:42]: It looks like something went wrong when you merged/rebased your branch onto master, this should not be changed by your PR I think.

File: doodler/components.py [hunk 2]
```diff
@@ -299,34 +302,43 @@ def __init__(self, **params):
...
-            raise ValueError(
-                f'The input image "{path}" has {nb_bands} channel(s), '
-                'only JPEG with 3 or 4 channels are supported.'
-            )
-        # Some JPEG files follow the CMYK model instead of RGB
-        if nb_bands == 4:
-            img = img.convert('RGB')
-        arr = np.array(img)
+        fn, ext = os.path.splitext(path)
+        print(ext)
```
Comment (maximlt) [inline doodler/components.py:323]: Stray print.

Comment (maximlt) [inline doodler/components.py:323]: Stray print.

File: doodler/components.py [hunk 3]
```diff
@@ -299,34 +302,43 @@ def __init__(self, **params):
...
+        if (ext.lower() == '.jpg') or (ext.lower() == '.jpeg'):
```
Comment (maximlt) [inline doodler/components.py:324]:
```suggestion
        if ext.lower() in ('.jpg', '.jpeg'):
```

Comment (maximlt) [inline doodler/components.py:324]:
```suggestion
        if ext.lower() in ('.jpg', '.jpeg'):
```

File: doodler/components.py [hunk 4]
```diff
@@ -299,34 +302,43 @@ def __init__(self, **params):
...
+            img = Image.open(path)
+            nbands = len(img.getbands())
+            # previously, code assumed 4 bands indicated CMYK, but band 4 can also be alpha, so:
+            if img.mode == 'CMYK':
+                img = img.convert('RGB')
+
+            # jpeg array is (nrow, ncol, nband)
+            arr = np.array(img)
+
+        elif (ext.lower() == '.tif') or (ext.lower == '.tiff'):
```
Comment (maximlt) [inline doodler/components.py:334]:
```suggestion
        if ext.lower() in ('.tif', '.tiff'):
```

Comment (maximlt) [inline doodler/components.py:334]:
```suggestion
        if ext.lower() in ('.tif', '.tiff'):
```

File: doodler/components.py [hunk 5]
```diff
@@ -299,34 +302,43 @@ def __init__(self, **params):
...
-        # Some JPEG files follow the CMYK model instead of RGB
-        if nb_bands == 4:
-            img = img.convert('RGB')
-        arr = np.array(img)
+        fn, ext = os.path.splitext(path)
+        print(ext)
+        if (ext.lower() == '.jpg') or (ext.lower() == '.jpeg'):
+            img = Image.open(path)
+            nbands = len(img.getbands())
+            # previously, code assumed 4 bands indicated CMYK, but band 4 can also be alpha, so:
```
Comment (maximlt) [inline doodler/components.py:327]: I remember checking how many bands could have a JPEG image and read that 4 bands meant CMYK (here: https://gdal.org/drivers/raster/jpeg.html#creation-options) but that source may be incomplete. I wasn't aware of the `mode` attribute which is indeed a better way anyway! You can remove the comment in the code, it's clear enough.

Comment (maximlt) [inline doodler/components.py:327]: I remember checking how many bands could have a JPEG image and read that 4 bands meant CMYK (here: https://gdal.org/drivers/raster/jpeg.html#creation-options) but that source may be incomplete. I wasn't aware of the `mode` attribute which is indeed a better way anyway! You can remove the comment in the code, it's clear enough.

**Review by maximlt — CHANGES_REQUESTED** [review]
Thanks for this PR @csherwood-usgs ! I left a few comments but I have a major question next to that. As far as I understood the doodler app doesn't have any geo dimension, the images that are ought to be labelled don't necessarily have to be aerial images, they could represent anything. With this PR you introduce `rioxarray` to read GeoTiff files, which will pull large dependencies like GDAL and PROJ (`rioxarray` depends on `pyproj` and `rasterio`). Is there a plan for the app to become more geo aware? Because if you just need to be able to open TIFF images, shouldn't PIL be enough? And if the app is meant to become geo aware then I guess a plan needs to laid out to see what to do with the geo metadata that comes with the GeoTIFF.

File: doodler/components.py (hunk 1 above)
Comment (csherwood-usgs) [reply doodler/components.py:42]: Agreed. Sorry.

File: doodler/components.py (hunk 2 above)
Comment (csherwood-usgs) [reply doodler/components.py:323]: Yes...should be removed.

File: doodler/components.py (hunk 3 above)
Comment (csherwood-usgs) [reply doodler/components.py:324]: Much nicer. Thanks.

File: doodler/components.py (hunk 4 above)
Comment (csherwood-usgs) [reply doodler/components.py:334]: Same. Thanks.

File: doodler/components.py (hunk 5 above)
Comment (csherwood-usgs) [reply doodler/components.py:327]: At least for many .tiff files, the fourth band can be alpha (transparency). But mode works well.

**Comment by csherwood-usgs** [comment]
I don't think we need the images as geo-aware, and you are correct...GDAL, xarray, rasterio are large and somewhat unstable packages that might be good to avoid. What we do want to do is read more bands...more than four. (As many as 12?). I will experiment with reading TIFFs with PIL...the key question is whether it can read >4 bands.

**Comment by maximlt** [comment]
There's also a library called tifffile which I've never used but that could be an alternative to PIL if it doesn't handle the kind of image you expect the app to support.

**Comment by csherwood-usgs** [comment]
I have confirmed that PIL cannot read my 6-band tif file. I will try tifffile. Thanks!

**Comment by csherwood-usgs** [comment]
My updated `crs_dev` branch now uses tifffile to read tif files. This seems to work fine on files with up to six bands (as far as I have tested). The next step is to pass those extra bands to the segmentation routine, while still passing the only the first three bands to the doodler panes. I envision keeping two arrays: a three-band image array (for doodling) and an n-band segmentation array (`seg+array` ) to pass to the segmentation routine. I am still trying to understand exactly where this happens, and which routines will be involved on tracking those two arrays.

**Comment by maximlt** [comment]
@csherwood-usgs thanks for the changes using tifffile and your explanations on the follow-up work. I'd suggest that if in the current state of this PR the app works as you expect then we should merge it.

As for adapting the core segmentation code, we just need to be careful not to have the code in this repo to diverge from the code in dash-doodler, both apps are meant to use the same code (which will happen when it is put into a package). I would suggest opening an issue in this repo that discusses the changes you want to make for the segmentation to support arrays with more than 3 bands, pinging Dan and others interested in this discussion.

**Comment by maximlt** [comment]
Superseded by https://github.com/pyviz-topics/holodoodler/pull/35

**Pull request closed without merging**